tests/lua/cmio-test.lua

```diff
@@ -0,0 +1,367 @@
+
+local remote_address
+local cleanup = {}
+
+-- Print help and exit
+local function help()
+    io.stderr:write(string.format(
+        [=[
+Usage:
+
+  %s <machine_type> [options]
+
+where options are:
+
+  --remote-address=<address>
+    run tests on a remote cartesi machine (when machine type is jsonrpc).
+
+<address> is one of the following formats:
+  <host>:<port>
+   unix:<path>
+
+<host> can be a host name, IPv4 or IPv6 address.
+]=],
+        arg[0]
+    ))
+    os.exit()
+end
+
+local options = {
+    {
+        "^%-%-h$",
+        function(all)
+            if not all then return false end
+            help()
+        end,
+    },
+    {
+        "^%-%-help$",
+        function(all)
+            if not all then return false end
+            help()
+        end,
+    },
+    {
+        "^%-%-remote%-address%=(.*)$",
+        function(o)
+            if not o or #o < 1 then return false end
+            remote_address = o
+            return true
+        end,
+    },
+    { ".*", function(all) error("unrecognized option " .. all) end },
+}
+
+-- Process command line options
+local arguments = {}
+for _, argument in ipairs({ ... }) do
+    if argument:sub(1, 1) == "-" then
+        for _, option in ipairs(options) do
+            if option[2](argument:match(option[1])) then break end
+        end
+    else
+        arguments[#arguments + 1] = argument
+    end
+end
+
+local machine_type = assert(arguments[1], "missing machine type")
+assert(machine_type == "local" or machine_type == "jsonrpc", "unknown machine type, should be 'local' or 'jsonrpc'")
+
+local protocol
+if machine_type == "jsonrpc" then
+    assert(remote_address ~= nil, "remote cartesi machine address is missing")
+    protocol = require("cartesi.jsonrpc")
+end
+
+local function connect()
+    local remote = protocol.stub(remote_address)
+    local version = assert(remote.get_version(), "could not connect to remote cartesi machine at " .. remote_address)
+    local shutdown = function() remote.shutdown() end
+    local mt = { __gc = function() pcall(shutdown) end }
+    setmetatable(cleanup, mt)
+    return remote, version
+end
+
+local remote
+
+-- There is no UINT64_MAX in Lua, so we have to use the signed representation
+local MAX_MCYCLE = -1
+local OUTPUTS_ROOT_HASH_SIZE = 32
+
+local function load_machine(name)
+    local runtime = {
+        concurrency = {
+            update_merkle_tree = 0,
+        },
+        skip_root_hash_check = true,
+        skip_root_hash_store = true,
+    }
+    if machine_type ~= "local" then
+        if not remote then remote = connect() end
+        return assert(remote.machine(MACHINES_DIR .. name, runtime))
+    else
+        return assert(cartesi.machine(MACHINES_DIR .. name, runtime))
+    end
+end
+
+local function set_yield_data(machine, reason, data)
+    local m16 = (1 << 16) - 1
+    local m32 = (1 << 32) - 1
+    machine:write_htif_fromhost_data(((reason & m16) << 32) | (data & m32))
```

Comment:
I am starting to think maybe we should have a specialized function to encode/decode yield data/reason. All this bit manipulation is a little fragile and being repeated in multiple places. I have noticed we did mistakes encoding/decoding this data field.

But not now, maybe in we could create an issue to do this later What you think @diegonehab ?
Could be something like `write_htif_yield_reason` same for `data`.

Reply:
Indeed, I forgot about it.